service/remotecfg/remotecfg.go

```diff
@@ -0,0 +1,318 @@
+			newConfigHash := getHash(b)
+			if s.getCfgHash() == newConfigHash {
+				level.Debug(s.opts.Logger).Log("msg", "skipping to the next polling loop")
+				continue
+			}
+
+			// We have a new configuration let's try to load it
+			err = s.parseAndLoad(b)
+			if err != nil {
+				level.Error(s.opts.Logger).Log("msg", "failed to load configuration from the API", "err", err)
+				continue
+			}
+
+			level.Info(s.opts.Logger).Log("msg", "new remote configuration loaded successfully")
+
+			// If successful, flush to disk and keep a copy.
+			err = os.WriteFile(s.dataPath, b, 0750)
+			if err != nil {
+				level.Error(s.opts.Logger).Log("msg", "failed to flush remote configuration contents the on-disk cache", "err", err)
+			}
+			s.setCfgHash(newConfigHash)
+		case <-ctx.Done():
+			s.ticker.Stop()
+			return nil
+		}
+	}
+}
+
+// Update implements [service.Service] and applies settings.
+func (s *Service) Update(newConfig any) error {
+	newArgs := newConfig.(Arguments)
+
+	// We either never set the block on the first place, or recently removed
+	// it. Make sure we stop everything gracefully before returning.
+	if newArgs.URL == "" {
+		s.mut.Lock()
+		s.ch = nil
+		s.ticker.Reset(math.MaxInt64)
+		s.asClient = noopClient{}
+		s.args.HTTPClientConfig = config.CloneDefaultHTTPClientConfig()
+		s.mut.Unlock()
+		return nil
+	}
+
+	s.mut.Lock()
+	hash, err := newArgs.Hash()
+	if err != nil {
+		return err
+	}
+	s.dataPath = filepath.Join(s.opts.StoragePath, ServiceName, hash)
+	s.ticker.Reset(newArgs.PollFrequency)
+	s.ch = s.ticker.C
+	// Update the HTTP client last since it might fail.
+	if !reflect.DeepEqual(s.args.HTTPClientConfig, newArgs.HTTPClientConfig) {
+		httpClient, err := commonconfig.NewClientFromConfig(*newArgs.HTTPClientConfig.Convert(), "remoteconfig")
+		if err != nil {
+			return err
+		}
+		s.asClient = agentv1connect.NewAgentServiceClient(
+			httpClient,
+			newArgs.URL,
+		)
+	}
+	s.args = newArgs // Update the args as the last step to avoid polluting any comparisons
+	s.mut.Unlock()
+
+	// If we've already called Run, then immediately trigger an API call with
+	// the updated Arguments, and/or fall back to the updated cache location.
+	if s.ctrl != nil && s.ctrl.Ready() {
+		s.fetch()
+	}
+
+	return nil
+}
+
+// fetch attempts to read configuration from the API and the local cache
+// and then parse/load in order of preference.
+func (s *Service) fetch() {
+	if !s.isEnabled() {
+		return
+	}
+
+	var b1, b2 []byte
+	b1, err := s.getAPIConfig()
+	b2, err2 := s.getCachedConfig()
+
+	if err != nil && err2 != nil {
+		level.Debug(s.opts.Logger).Log("msg", "failed to get any configuration during startup", "apiErr", err, "cacheErr", err2)
+		return
+	}
+
+	err = s.parseAndLoad(b1)
+	if err != nil && err2 == nil {
+		err = s.parseAndLoad(b2)
+	}
+
+	if err != nil {
+		level.Error(s.opts.Logger).Log("msg", "failed to load remote cfg during startup", "err", err)
+	}
```

Comment:
I find the logic a bit hard to follow, especially because we are keeping track of two errors. I would like to suggest the following approach (not tested):

```
b, err := s.getAPIConfig()
if err != nil {
  err = s.parseAndLoad(b)
  if err == nil {
    return
  }
}
level.Error(s.opts.Logger).Log("msg", "failed to load remote cfg during startup, falling back to cache", "err", err)
b, err = s.getCachedConfig()
if err != nil {
    err = s.parseAndLoad(b)
    if err == nil {
      return
   }
}
level.Error(s.opts.Logger).Log("msg", "failed to load cached remote cfg during startup", "err", err)
```

Reply:
Readability matters, right? 😅

The reason I tried the current way is that we're all used to `if err != nil` conditions in 90% of the code we read/write, so it's be easy to skim over and misunderstand what's happening here, but I guess the added implicitness isn't worth it.

Reply:
The reason I tried the current way is that we're all used to `if err != nil` conditions in 90% of the code we read/write, so it's be easy to skim over and misunderstand what's happening here, but I guess the added implicitness isn't worth it.

Reply:
agreed that the `if err == nil` is a bit special but imo it's one of the rare cases where it makes things easier to follow

Reply:
In your above example, _all_ conditions have to be `err == nil`, so it's indeed hard to write/read even as a small example.

In any case, it doesn't matter for how the feature works, so I'm ok with changing it.